mobilitypath_publisher/src/mobilitypath_publisher.cpp

```diff
@@ -0,0 +1,144 @@
+    {
+        current_pose_ = msg;
+    }
+
+    void MobilityPathPublication::bsm_cb(const cav_msgs::BSMConstPtr& msg)
+    {
+        bsm_core_ = msg->core_data;
+        
+    }
+    
+// @SONAR_START@
+    cav_msgs::MobilityPath MobilityPathPublication::mobilityPathMessageGenerator(const cav_msgs::TrajectoryPlan &trajectory_plan, geometry_msgs::TransformStamped tf)
+    {
+        cav_msgs::MobilityPath mobility_path_msg;
+        uint64_t millisecs =trajectory_plan.header.stamp.toNSec()/1000;
+        mobility_path_msg.header = composeMobilityHeader(millisecs);
+        
+        cav_msgs::Trajectory mob_path_traj = TrajectoryPlantoTrajectory(trajectory_plan.trajectory_points, tf);
+        mobility_path_msg.trajectory = mob_path_traj;
+
+        return mobility_path_msg;
+    }
+
+    cav_msgs::MobilityHeader MobilityPathPublication::composeMobilityHeader(uint64_t time){
+        cav_msgs::MobilityHeader header;
+        header.sender_id = sender_id;
+        header.recipient_id = recipient_id;
+        header.sender_bsm_id = sender_bsm_id = bsmIDtoString(bsm_core_);// sender_bsm_id;
+        // random GUID that identifies this particular plan for future reference
+        header.plan_id = boost::uuids::to_string(boost::uuids::random_generator()());
+        header.timestamp = time;//ros::Time::now().toNSec()/1000; //time in millisecond
+        
+        return header;
+    }
+
+    cav_msgs::Trajectory MobilityPathPublication::TrajectoryPlantoTrajectory(std::vector<cav_msgs::TrajectoryPlanPoint> traj_points, geometry_msgs::TransformStamped tf){
+        cav_msgs::Trajectory traj;
+        cav_msgs::LocationECEF ecef_location = TrajectoryPointtoECEF(traj_points[0], tf);
+
+        if (traj_points.size()<2){
+            ROS_WARN("Received Trajectory Plan is too small");
+            traj.offsets = {};
+        }
+        else{
+            for (size_t i=1; i<traj_points.size(); i++){
+                
+                    cav_msgs::LocationOffsetECEF offset;
+                    cav_msgs::LocationECEF new_point = TrajectoryPointtoECEF(traj_points[i], tf);
+                    offset.offset_x = new_point.ecef_x - ecef_location.ecef_x;
+                    offset.offset_y = new_point.ecef_y - ecef_location.ecef_y;
+                    offset.offset_z = new_point.ecef_z - ecef_location.ecef_z;
+                    traj.offsets.push_back(offset);
+                
+            }
+        }
+        
+        
+        traj.location = ecef_location;
+        return traj;
+    }
+
+
+
+    cav_msgs::LocationECEF MobilityPathPublication::TrajectoryPointtoECEF(cav_msgs::TrajectoryPlanPoint traj_point, geometry_msgs::TransformStamped tf){
+        cav_msgs::LocationECEF ecef_point;    
+
+        ecef_point.ecef_x = traj_point.x * tf.transform.translation.x;
+        ecef_point.ecef_y = traj_point.x * tf.transform.translation.y;
+        ecef_point.ecef_z = traj_point.x * tf.transform.translation.z;
```

Comment:
I am not sure how traj_point works but should these be traj_point.x traj_point.y traj_point.z?